runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/FusedPipeline.java

```diff
@@ -19,54 +19,84 @@
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
 import org.apache.beam.model.pipeline.v1.RunnerApi.Components;
-import org.apache.beam.model.pipeline.v1.RunnerApi.Components.Builder;
 import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
 import org.apache.beam.model.pipeline.v1.RunnerApi.Pipeline;
 import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
 
-/**
- * A {@link Pipeline} which has been separated into collections of executable components.
- */
+/** A {@link Pipeline} which has been separated into collections of executable components. */
 @AutoValue
 public abstract class FusedPipeline {
   static FusedPipeline of(
       Set<ExecutableStage> environmentalStages, Set<PTransformNode> runnerStages) {
     return new AutoValue_FusedPipeline(environmentalStages, runnerStages);
   }
 
-  /**
-   * The {@link ExecutableStage executable stages} that are executed by SDK harnesses.
-   */
+  /** The {@link ExecutableStage executable stages} that are executed by SDK harnesses. */
   public abstract Set<ExecutableStage> getFusedStages();
 
-  /**
-   * The {@link PTransform PTransforms} that a runner is responsible for executing.
-   */
+  /** The {@link PTransform PTransforms} that a runner is responsible for executing. */
   public abstract Set<PTransformNode> getRunnerExecutedTransforms();
 
+  public RunnerApi.Pipeline toPipeline(Components initialComponents) {
+    Components executableComponents =
+        initialComponents
+            .toBuilder()
+            .clearTransforms()
+            .putAllTransforms(getTopLevelTransforms(initialComponents))
+            .build();
+    List<String> rootTransformIds =
+        StreamSupport.stream(
+                QueryablePipeline.forComponents(executableComponents)
+                    .getTopologicallyOrderedTransforms()
+                    .spliterator(),
+                false)
+            .map(PTransformNode::getId)
+            .collect(Collectors.toList());
+    return Pipeline.newBuilder()
+        .setComponents(executableComponents.toBuilder().putAllTransforms(getFusedTransforms()))
+        .addAllRootTransformIds(rootTransformIds)
+        .build();
+  }
+
   /**
    * Return a {@link Components} like the {@code base} components, but with the only transforms
    * equal to this fused pipeline.
    *
-   * <p>The only composites will be the stages returned by {@link #getFusedStages()}.
+   * <p>The only composites will be the stages returned by {@link #getFusedStages()}, and the only
+   * primitives will be the {@link PTransformNode transforms} returned by {@link
+   * #getRunnerExecutedTransforms()}.
    */
-  public Components asComponents(Components base) {
-    Builder newComponents = base.toBuilder().clearTransforms();
+  private Map<String, PTransform> getTopLevelTransforms(Components base) {
```

Comment:
getRunnerExecutedPrimitiveTransforms?

Reply:
`getExecutableTransforms` is what I am calling this now.

All of the transforms are primitives, but there are non-executable transforms (within a stage) that get merged into the components as well.